### PR TITLE
changes in aligning cloud storage applications and events to take into account more than one app per connector, authentication and config values

### PR DESCRIPTION
**Pull request by yourgirlcodes**

### CODE REVIEW

File: boilerplate/index.ts [hunk 1]
```diff
@@ -498,11 +498,24 @@ export function createBoilerplate(sdk: CDP) {
...
+                                    ...boilerplateEvent,
+                                    schema: JSON.stringify(boilerplateEvent.schema),
+                                    purposeIds: boilerplateEvent.purposeIds
+                                });
+                            } else {
+                                return bOps.applications.for(remoteCloudStorageApplication.id).dataevents.create({
+                                    ...boilerplateEvent,
+                                    schema: JSON.stringify(boilerplateEvent.schema),
+                                    purposeIds: boilerplateEvent.purposeIds,
+                                    configValues: {
```
Comment (iBaryo) [inline boilerplate/index.ts:512]: why is this required?

File: boilerplate/index.ts [hunk 2]
```diff
@@ -546,7 +559,7 @@ export function createBoilerplate(sdk: CDP) {
                                 return _(bpMappings).differenceWith(rMappings, _.isEqual).isEmpty();
                             };
 
-                            if (remoteMappings.length < 1) {
+                            if (remoteMappings.length < 1 || !remoteMappings) {
```
Comment (iBaryo) [inline boilerplate/index.ts:562]: condition order doesn't make sense. if `!remoteMappings` then the first part of the condition will throw.

File: boilerplate/index.ts [hunk 3]
```diff
@@ -567,19 +580,22 @@ export function createBoilerplate(sdk: CDP) {
...
                         const remoteConnectors = await sdk.api.workspaces.for(config.workspaceId).applibrary.getAll({includePublic: true});
 
-                        const boilerplateConnectorTypes: CSType[] = [ 'AWS S3', 'Microsoft Azure Blob', 'Google Cloud Storage', 'SFTP']
+                        const boilerplateConnectorTypes: CSType[] = ['AWS S3', 'Microsoft Azure Blob', 'Google Cloud Storage', 'SFTP']
 
                         // get remote connectors that are Cloud Storage connectors
-                        const remoteCloudStorageConnectors = remoteConnectors &&
-                            (remoteConnectors?.filter(connector => connector.type === 'CloudStorage' && (boilerplateConnectorTypes.includes(connector.name as CSType))));
+                        const remoteCloudStorageConnectors =
+                            JSON.parse(JSON.stringify(keepUnique(remoteConnectors, conn => conn.name))).filter(connector => connector.type === 'CloudStorage' && (boilerplateConnectorTypes.includes(connector.name as CSType)));
```
Comment (iBaryo) [inline boilerplate/index.ts:587]: why cloning?

File: boilerplate/index.ts [hunk 4]
```diff
@@ -567,19 +580,22 @@ export function createBoilerplate(sdk: CDP) {
...
 
+                        function keepUnique(data, key) {
```
Comment (iBaryo) [inline boilerplate/index.ts:589]: better move to utils

File: boilerplate/index.ts [hunk 5]
```diff
@@ -596,8 +612,26 @@ export function createBoilerplate(sdk: CDP) {
                                     configValues: boilerplateCloudStorageApplication.configValues,
                                     enabled: true
                                 };
-                                remoteCloudStorageApplication = await bOps.applications.create(cloudStoragePayload).then(app => bOps.applications.for(app.id).get());
+                                remoteCloudStorageApplication = await bOps.applications.create(cloudStoragePayload).then(app => {
+                                    if (app.name !== "Microsoft Azure Blob") {
+                                        bOps.applications.for(app.id).auth.create({
```
Comment (iBaryo) [inline boilerplate/index.ts:617]: why is setting the application auth is required for the boilerplate?

File: boilerplate/index.ts (hunk 1 above)
Comment (yourgirlcodes) [reply boilerplate/index.ts:512]: use application configSchema

File: boilerplate/index.ts (hunk 5 above)
Comment (yourgirlcodes) [reply boilerplate/index.ts:617]: email Michael

File: boilerplate/index.ts (hunk 1 above)
Comment (yourgirlcodes) [reply boilerplate/index.ts]: unfortunately we cant use the config schema since it holds many irrelevant fields. This would mean sorting through the fields, which is more complex than explicitly stating what we need

File: boilerplate/index.ts [hunk 6]
```diff
@@ -94,47 +94,47 @@ export function createBoilerplate(sdk: CDP) {
...
                                 alignActivityPromise = !fieldDiffs.length ?
                                     Promise.resolve(activitySchema)
-                                    : bOps.ucpschemas.for(activitySchema.id).update({
-                                        enabled: true,
+                                    : bOps.customerschemas.for(activitySchema.id).update({
                                         name: activity,
-                                        schema: JSON.stringify({
-                                            ...remoteActivitySchema,
+                                        schema: {
+                                            ...remoteActivitySchema as Object,
```
Comment (iBaryo) [inline boilerplate/index.ts:132]: if the schema now always returns as an object, could you also please change it in the sdk [then you can remove this casting]

File: boilerplate/index.ts [hunk 7]
```diff
@@ -497,12 +493,27 @@ export function createBoilerplate(sdk: CDP) {
                             });
                         }
 
+
                         function createCloudStorageEvent(boilerplateEvent, remoteCloudStorageApplication) {
+                            // const conf = {};
+                            // for (const key of Object.keys(remoteCloudStorageApplication.configSchema.properties)) {
+                            //     conf[key] = 'value';
+                            // }
+
```
Comment (iBaryo) [inline boilerplate/index.ts:502]: remove commented out code

File: boilerplate/index.ts [hunk 8]
```diff
@@ -497,12 +493,27 @@ export function createBoilerplate(sdk: CDP) {
...
+                                purposeIds: boilerplateEvent.purposeIds,
+                                configValues: remoteCloudStorageApplication.name === 'Microsoft Azure Blob' ?
+                                    {
+                                        // these need to be explicitly defined and cannot be taken from the connector's or application's config schema since there is a ton of irrelevant info there
+                                        readContainer: "any container",
+                                        readFileNameRegex: null,
+                                        readFormat: null,
+                                    }
+                                    :
+                                    boilerplateEvent.configValues
```
Comment (iBaryo) [inline boilerplate/index.ts:515]: just a flavor comment we talked about before: short `if`, longer `else`  [means to swap if/else]

File: boilerplate/index.ts [hunk 9]
```diff
@@ -546,7 +557,7 @@ export function createBoilerplate(sdk: CDP) {
                                 return _(bpMappings).differenceWith(rMappings, _.isEqual).isEmpty();
                             };
 
-                            if (remoteMappings.length < 1) {
+                            if (!remoteMappings || remoteMappings.length < 1) {
```
Comment (iBaryo) [inline boilerplate/index.ts:560]: `remoteMappings?.length < 1`

File: boilerplate/index.ts [hunk 10]
```diff
@@ -564,49 +575,51 @@ export function createBoilerplate(sdk: CDP) {
...
+                            keepUnique(remoteConnectors, conn => conn.name).filter((connector: Connector) => connector.type === 'CloudStorage' && connector.enabled && (boilerplateConnectorTypes.includes(connector.name as CSType)));
+                        await Promise.all(remoteCloudStorageConnectors.map(async (connector: Connector) => {
                             // get the corresponding cloud storage application
 
-                            //get cloud storage applications with the same type as connector // npt working
-                            let remoteCloudStorageApplication = allCloudStorageRemoteApplications?.find(application => (application['originConnectorId'] === connector.id))
-
-                            let remoteCloudStorageApplicationId: ApplicationId = remoteCloudStorageApplication?.id;
+                            // get cloud storage applications created from connector
+                            let remoteCloudStorageApplication = allCloudStorageRemoteApplications?.find(application => (application['originConnectorId'] === connector.id || application.name === connector.name))
```
Comment (iBaryo) [inline boilerplate/index.ts:589]: `originConnectorId` added to `Application` definition.
also - i think you can drop the part in the condition regarding the name.

Comment (yourgirlcodes) [reply boilerplate/index.ts:589]: regarding originConnectorId --> see slack messages

File: boilerplate/index.ts [hunk 11]
```diff
@@ -497,12 +493,23 @@ export function createBoilerplate(sdk: CDP) {
...
+
                         function createCloudStorageEvent(boilerplateEvent, remoteCloudStorageApplication) {
+
                             return bOps.applications.for(remoteCloudStorageApplication.id).dataevents.create({
                                 ...boilerplateEvent,
                                 schema: JSON.stringify(boilerplateEvent.schema),
-                                purposeIds: boilerplateEvent.purposeIds
-                            });
+                                purposeIds: boilerplateEvent.purposeIds,
+                                configValues: remoteCloudStorageApplication.name === 'Microsoft Azure Blob' ?
```
Comment (iBaryo) [inline boilerplate/index.ts:503]: need to be `!==` no?

File: boilerplate/index.ts [hunk 12]
```diff
@@ -546,7 +553,7 @@ export function createBoilerplate(sdk: CDP) {
                                 return _(bpMappings).differenceWith(rMappings, _.isEqual).isEmpty();
                             };
 
-                            if (remoteMappings.length < 1) {
+                            if (!remoteMappings || remoteMappings?.length < 1) {
```
Comment (iBaryo) [inline boilerplate/index.ts:556]: suggestion [not mandatory]:
`if ((remoteMapping?.length ?? 0) < 1)`
